### PR TITLE
[FLOC-3607] Construct actions uniformly.

### PR DESCRIPTION
**Pull request by tomprince**
https://clusterhq.atlassian.net/browse/FLOC-3607

### CODE REVIEW

File: flocker/node/agents/test/test_blockdevice.py [hunk 1]
```diff
@@ -2201,6 +2201,53 @@ def test_match_configuration_to_state_of_datasets(self):
 
         self.assertEqual(expected_changes, actual_changes)
 
+    def test_dataset_with_metadata(self):
+        """
+        XXX
```
Comment (jml) [inline flocker/node/agents/test/test_blockdevice.py]: Please add a real docstring here :)

File: flocker/node/agents/test/test_blockdevice.py [hunk 2]
```diff
@@ -2201,6 +2201,53 @@ def test_match_configuration_to_state_of_datasets(self):
...
+        )
+
+        configuration = Deployment(
+            nodes={
+                Node(
+                    uuid=node_id,
+                    manifestations={
+                        dataset_id: Manifestation(
+                            dataset=requested_dataset,
+                            primary=True
```
Comment (jml) [inline flocker/node/agents/test/test_blockdevice.py]: Nit: trailing comma please

File: flocker/node/agents/blockdevice.py [hunk 3]
```diff
@@ -114,6 +114,21 @@ class DiscoveredDataset(PClass):
     )
 
 
+class DesiredDataset(PClass):
+    """
+    Dataset as requested by configuration (and applications).
```
Comment (jml) [inline flocker/node/agents/blockdevice.py]: I don't really understand why "(and applications)" is in parentheses. Will they use `DesiredDataset` in a different way somehow? 

Could you please either remove the parens or expand on the difference in the docstring.

File: flocker/node/agents/blockdevice.py [hunk 4]
```diff
@@ -1453,6 +1510,22 @@ def _mountpath_for_dataset_id(self, dataset_id):
         """
         return self.mountroot.child(dataset_id.encode("ascii"))
 
+    def _calculate_desired_for_manifestation(self, manifestation):
```
Comment (jml) [inline flocker/node/agents/blockdevice.py]: Please add a docstring.

File: flocker/node/agents/blockdevice.py [hunk 5]
```diff
@@ -410,6 +425,13 @@ class DestroyBlockDeviceDataset(PRecord):
     dataset_id = field(type=UUID, mandatory=True)
     blockdevice_id = field(type=unicode, mandatory=True)
 
+    @classmethod
+    def from_state_and_config(cls, discovered_dataset, desired_dataset):
```
Comment (jml) [inline flocker/node/agents/blockdevice.py]: Please add docstrings for these methods.

Comment (jml) [inline flocker/node/agents/blockdevice.py]: :question: Should there be an interface for `from_state_and_config` that is provided by these classes?

**Comment by jml** [comment]
Thanks Tom, please address comments, make peace w/ the test suite & merge.

**Comment by tomprince** [comment]
I'm going to merge this as-is (with failing rackspace builds). The only changes since having succesful builds is merging forward and test/docstring changes.

**Pull request merged**